rectools/models/sasrec.py

```diff
@@ -1071,41 +1018,191 @@ def _calc_softmax_loss(cls, logits: torch.Tensor, y: torch.Tensor, w: torch.Tens
+        lr: float,
+        gbce_t: float,
+        n_actual_items: int,
+        **kwargs: tp.Any,
+    ):
+        super().__init__()
+        self.lr = lr
+        self.torch_model = torch_model
+        self.gbce_t = gbce_t
+        self.item_embs: torch.Tensor
+        if isinstance(loss, SessionEncoderHeadBase):
+            self.session_encoder_head = loss
+        else:
+            self.session_encoder_head = self._init_head(loss, n_actual_items)
+
+    def _init_head(self, loss: LossName, n_actual_items: int) -> SessionEncoderHeadBase:
+
+        if loss == "softmax":
+            return DotProductSoftmaxHead()
+        if loss == "BCE":
+            return DotProductBCEHead()
+        if loss == "gBCE":
+            return DotProductGBCEHead(self.gbce_t, n_actual_items)
+        raise ValueError(f"Unknown loss {loss}.")
+
+    def training_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
+        """Training step."""
+        raise NotImplementedError()
+
+    def predict_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
+        """Prediction step."""
+        raise NotImplementedError()
+
+    def configure_optimizers(self) -> OptimizerLRScheduler:
+        """Choose what optimizers and learning-rate schedulers to use in optimization"""
+        raise NotImplementedError()
+
+
+class SessionEncoderLightningModule(SessionEncoderLightningModuleBase):
+    """Lightning module to train transformer based models."""
+
+    def on_train_start(self) -> None:
+        """Initialize parameters with values from Xavier normal distribution."""
+        # TODO: init padding embedding with zeros
+        self._xavier_normal_init()
+
+    def training_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
+        """Forward pass and loss calculation."""
+        item_embs, session_embs = self.torch_model(batch["x"])
+        return self.session_encoder_head(batch, item_embs, session_embs)
 
     def on_train_end(self) -> None:
-        """Save item embeddings"""
+        """Save item embeddings for inference after training end."""
         self.eval()
         self.item_embs = self.torch_model.item_model.get_all_embeddings()
 
-    def predict_step(self, batch: Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
-        """
-        Prediction step.
-        Encode user sessions.
-        """
+    def predict_step(self, batch: tp.Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
+        """Prediction step. Encode user sessions. Get only last position hidden state from each user session."""
         encoded_sessions = self.torch_model.encode_sessions(batch["x"], self.item_embs.to(self.device))[:, -1, :]
         return encoded_sessions
 
+    def configure_optimizers(self) -> torch.optim.Adam:
+        """Choose what optimizers and learning-rate schedulers to use in optimization"""
+        adam_betas = (0.9, 0.98)
```

Comment:
Is it going to be a hardcoded value?

Reply:
If you plan to change it later, please add TODO here

Reply:
It's gonna be hardcoded. Can't move all of the params to user, We have too many small things here and there.